### PR TITLE
[ValueTracking] Compute known bits from recursive select/phi

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch is inspired by https://github.com/llvm/llvm-project/pull/113686. I found that it removes a lot of unnecessary "and X, 1" in some applications that represent boolean values with int.

### CODE REVIEW

**Comment by goldsteinn** [comment]
This looks fine although I have two notes.

1) since we are sure to be breaking the recursion here we can probably do a full depth computeKnownBits below.

2) it would be nice if we could also use the selects compare context.

**Comment by dtcxzyw** [comment]
> it would be nice if we could also use the selects compare context.

Looks like it breaks the stage2 build.

**Comment by dtcxzyw** [comment]
> it would be nice if we could also use the selects compare context.

It doesn't improve anything (with +0.01% compile-time overhead). Should I drop it?

**Comment by goldsteinn** [comment]
> > it would be nice if we could also use the selects compare context.
> 
> It doesn't improve anything (with +0.01% compile-time overhead). Should I drop it?

If it doesn't improve anything yeah, although IMO leave a comment explaining as much.

**Review by goldsteinn — APPROVED** [review]
LGTM like this or if you want to revert to prior version.

**Comment by goldsteinn** [comment]
Are you going to update to handle https://github.com/llvm/llvm-project/pull/114008.
If not I can post a patch after you merge this.

**Comment by dtcxzyw** [comment]
> Are you going to update to handle #114008. If not I can post a patch after you merge this.

No, feel free to do this :)

File: llvm/lib/Analysis/ValueTracking.cpp
Comment (nikic) [inline llvm/lib/Analysis/ValueTracking.cpp:1581]: Forwarding the question from https://github.com/llvm/llvm-project/pull/114689#pullrequestreview-2411822612 because the issue was introduced here: Why do we use the full depth for the select case?

As far as I can tell, a) there may be multiple incoming values in this form, resulting in a fan-out problem and b) we are not actually guaranteed that this breaks recursion (because the other select operand may also be based on the phi *indirectly*) so the fan-out may be recursive as well.

Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:1581]: Answered in https://github.com/llvm/llvm-project/pull/114689#issuecomment-2453461016, lets keep the conversion in one place.

**Pull request merged**
